sieve-server/obs_gap_server.go

```diff
@@ -113,56 +101,49 @@ func (s *obsGapServer) NotifyObsGapAfterIndexerWrite(request *sieve.NotifyObsGap
 	}
 	// If we are inside pausing, then we check for target event which can cancel the crucial one
-	s.mutex.RLock()
-	pausingReconcile := s.pausingReconcile
-	s.mutex.RUnlock()
-	log.Println("NotifyObsGapAfterIndexerWrite", pausingReconcile, "pausedReconcileCnt", s.pausedReconcileCnt)
-	if pausingReconcile {
-		if request.OperationType == "Deleted" {
-			log.Printf("[sieve] we met the later cancel event %s, reconcile is resumed, paused cnt: %d\n", request.OperationType, s.pausedReconcileCnt)
+	log.Println("NotifyObsGapAfterIndexerWrite", s.pausingReconcile)
+	if s.pausingReconcile {
+		if request.OperationType == "Deleted" || conflictingEventAsMap(s.diffCurEvent, currentEvent) {
+			// TODO: we should also consider the corner case where s.diffCurEvent == {}
+			log.Printf("[sieve] we met the cancel event %s, reconcile is resumed\n", request.OperationType)
 			log.Println("NotifyObsGapAfterIndexerWrite", request.OperationType, request.ResourceType, request.Object)
-			s.mutex.Lock()
 			s.pausingReconcile = false
-			s.cond.Broadcast()
-			s.mutex.Unlock()
+			s.reconcilingMutex.Unlock()
 			finishObsGapInjection()
-		} else {
-			// TODO: we should also consider the corner case where s.diffCurEvent == {}
-			if conflictingEventAsMap(s.diffCurEvent, currentEvent) {
-				log.Printf("[sieve] we met the later cancel event %s, reconcile is resumed, paused cnt: %d\n", request.OperationType, s.pausedReconcileCnt)
-				log.Println("NotifyObsGapAfterIndexerWrite", request.OperationType, request.ResourceType, request.Object)
-				s.mutex.Lock()
-				s.pausingReconcile = false
-				s.cond.Broadcast()
-				s.mutex.Unlock()
-				finishObsGapInjection()
-			}
 		}
 	}
 	*response = sieve.Response{Ok: true}
 	return nil
 }
 
-func (s *obsGapServer) NotifyObsGapBeforeReconcile(request *sieve.NotifyObsGapBeforeReconcileRequest, response *sieve.Response) error {
-	s.reconcilingMutex.Lock()
-	recID := request.ControllerName
-	s.mutex.Lock()
-	log.Println("NotifyObsGapBeforeReconcile[0/1]", recID, s.pausingReconcile)
-	if s.pausingReconcile {
-		atomic.AddInt32(&s.pausedReconcileCnt, 1)
-	}
-	for s.pausingReconcile {
-		s.cond.Wait()
+func (s *obsGapServer) NotifyObsGapBeforeInformerCacheRead(request *sieve.NotifyObsGapBeforeInformerCacheReadRequest, response *sieve.Response) error {
+	if request.OperationType == "Get" {
+		if !(request.ResourceType == s.ceRtype && request.Name == s.ceName && request.Namespace == s.ceNamespace) {
+			log.Fatalf("encounter unexpected object: %s %s %s", request.ResourceType, request.Namespace, request.Name)
+		}
+	} else {
+		if !(request.ResourceType == s.ceRtype+"list") {
+			log.Fatalf("encounter unexpected object: %s", request.ResourceType)
+		}
 	}
-	log.Println("NotifyObsGapBeforeReconcile[1/1]", recID, s.pausingReconcile)
-	s.mutex.Unlock()
+	log.Println("NotifyObsGapBeforeInformerCacheRead[0/1]", s.pausingReconcile)
+	s.reconcilingMutex.Lock()
```

Comment:
Just one question here:
Or originally the critical session is reconcile wise, but now we make it per cache read/list wise. In that case, the reconcile goroutine may be paused inside a read/list a long while (waiting for cancel event). And after the pausing, the cluster state which the operator read before may not be up-to-date, but they will still try to use that state heavily (maybe for some resource's later update), not sure whether that can bring additional inconsistency for our testing.

Reply:
It is a great point! Let me put it into the comment. We can revisit this if we later encounter such case.